### ISSUE TITLE
backup with restic failed because of "restic repository is not ready"

### ISSUE THREAD

**Issue by qd19zzx** [vmware-tanzu/velero#3134]
Hi, I am using velero v 1.52 and alibaba velero-plugin. When I backup with restic, the velero pod reports an error like below: 
 
error="restic repository is not ready: restic repository prefix (resticRepoPrefix) not specified in backup storage location's config"

And when I set the "resticRepoPrefix" in the backupstoragelocation.config.resticRepoPrefix, the velero pods said that the resticRepoPrefix is not a valid key. 

So I want to know how to backup with restic? If is need to set the "resticRepoPrefix", where should I to set that key. Thanks.

**Comment by nrb** [comment]
Can you please provide what the `resticRepoPrefix` string is that you're using?

**Comment by qd19zzx** [comment]
Hi @nrb, I'm also not sure what is the  "resticRepoPrefix".
When I backup with restic, it always failed for the pv backup. And from the log of velero pod, it said 'error="restic repository is not ready: restic repository prefix (resticRepoPrefix) not specified in backup storage location's config"'

Here is my backup-location yaml file
``` apiVersion: velero.io/v1
kind: BackupStorageLocation
metadata:
  annotations:
    helm.sh/hook: post-install,post-upgrade
    helm.sh/hook-delete-policy: before-hook-creation
  creationTimestamp: "2020-12-07T05:39:45Z"
  generation: 1
  labels:
    app.kubernetes.io/instance: wiggly-chimp
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/name: velero
    helm.sh/chart: velero-2.14.1
  name: default
  namespace: velero
  resourceVersion: "47142790"
  selfLink: /apis/velero.io/v1/namespaces/velero/backupstoragelocations/default
  uid: 349c8881-3349-4921-9089-a78f25572739
spec:
  config:
    region: cn-beijing
  objectStorage:
    bucket: qd19zzx
    prefix: backup
  provider: alibabacloud
status:
  lastSyncedTime: "2020-12-08T00:38:17Z"
  lastValidationTime: "2020-12-08T00:38:17Z"
  phase: Available

Regards the error, It seems I should set the resticRepoPrefix in the config field of backup-location. But when I did that, the log of velero pods return another error, that the "resticRepoPrefix is not a valid key". I have tried with aws provider and alibaba provider. Both cannot set the "resticRepoPrefix".

**Comment by zubron** [comment]
Thanks for the additional info, @qd19zzx. There are a few things going on here, which I'll outline below.

First, the error that you are seeing that `resticRepoPrefix is not a valid key` is an error coming from the Alibaba plugin. Velero provides a utility function for plugins to validate the keys that are passed through via the `config` map. In this case, the Alibaba plugin [does not include `resticRepoPrefix` in its list of valid config keys](https://github.com/AliyunContainerService/velero-plugin/blob/master/velero-plugin-for-alibabacloud/object_store.go#L95-L97). The value for `resticRepoPrefix` should be the cloud storage URL where all namespace restic repos will be created (velero creates one restic repo _per namespace_). Velero handles the creation of this prefix for [Amazon, Azure, and GCP](https://github.com/vmware-tanzu/velero/blob/main/pkg/restic/config.go#L67-L89). For example, if backing up 2 namespaces (`ns1` and `ns2`) using restic on AWS, the `resticRepoPrefix` would be something like `s3:s3-us-west-2.amazonaws.com/bucket/restic` but the full restic repo path for `ns1` would be `s3:s3-us-west-2.amazonaws.com/bucket/restic/ns1`. We have opened #3136 to update our documentation to describe this config field.

Second, we initially thought Alibaba Cloud OSS was not S3 compatible and would require using `rclone` which Velero does not support. However, after looking through some issues in restic project, I came across this comment: https://github.com/restic/restic/issues/1354#issuecomment-453763202.
It looks like the Alibaba _is_ S3 compatible but [requires additional flags](https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#alibaba-cloud-aliyun-object-storage-system-oss) to be set when calling restic commands. Adding additional flags is not something that Velero currently supports so we would need to make changes to support this.

In summary, there are two things that would need to change:
- The Alibaba plugin will need to be updated to allow the `resticRepoPrefix` config key
- Velero would need to add support for setting additional flags in restic commands

**Comment by qd19zzx** [comment]
@zubron Thanks, appreciate for your detail explanation
